### PR TITLE
[CIS-2235] Properly arrange Authentication and Connection logic

### PR DESCRIPTION
**Pull request by polqf**
### 🔗 Issue Links

https://stream-io.atlassian.net/browse/CIS-2235

### 🎯 Goal

_Describe why we are making this change._

### 📝 Summary

_Provide bullet points with the most important changes in the codebase._

### 🛠 Implementation

_Provide a detailed description of the implementation and explain your decisions if you find them relevant._

### 🎨 Showcase

_Add relevant screenshots and/or videos/gifs to easily see what this PR changes, if applicable._

| Before | After |
| ------ | ----- |
|  img   |  img  |

### 🧪 Manual Testing Notes

_Explain how this change can be tested manually, if applicable._

### ☑️ Contributor Checklist

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

_Provide a funny gif or image that relates to your work on this pull request. (Optional)_

### CODE REVIEW

File: Sources/StreamChat/Repositories/ConnectionRepository.swift
Comment (martinmitrevski) [inline Sources/StreamChat/Repositories/ConnectionRepository.swift:54]: We should also document the `userInfo` here.

File: Sources/StreamChat/ChatClient.swift
Comment (martinmitrevski) [inline Sources/StreamChat/ChatClient.swift:461]: Is this comment still relevant?

File: Sources/StreamChat/ChatClient.swift
Comment (martinmitrevski) [inline Sources/StreamChat/ChatClient.swift:237]: Great that we're getting rid of these - now they're in the correct place and no `@Atomic` is needed. 👍

File: Sources/StreamChat/Repositories/ConnectionRepository.swift
Comment (martinmitrevski) [inline Sources/StreamChat/Repositories/ConnectionRepository.swift:14]: Not in the scope of this PR, but food for thought: since we will be adding more thread-safe repos in the future, maybe we should think about abstracting this logic for getters and setters, since it will repeat itself?

**Review by martinmitrevski — COMMENTED** [review]
Looks really nice @polqf, much cleaner and in line with the direction we want to go in 👏 

As already discussed, we should merge this after our next release. Also, @alteral's token-related e2e tests would need to be run before merging. We might even do a common session to do some QA, if it makes sense.

Additionally, we need tests and see why the checks are failing.

File: Sources/StreamChat/Repositories/ConnectionRepository.swift
Comment (polqf) [reply Sources/StreamChat/Repositories/ConnectionRepository.swift:14]: I did not do that because it requires us to make that wrapping class as public in some cases...
But might be trying to find a workaround on that soon

File: Sources/StreamChat/Repositories/AuthenticationRepository.swift
Comment (martinmitrevski) [inline Sources/StreamChat/Repositories/AuthenticationRepository.swift:173]: The handlings for both `firstConnection` and `newToken` are the same. Maybe we can put them in the same case (or extract these lines in a method)?

File: Sources/StreamChat/Repositories/AuthenticationRepository.swift
Comment (martinmitrevski) [inline Sources/StreamChat/Repositories/AuthenticationRepository.swift:197]: I think I've commented somewhere about the order of these two - it's clearer if we first clear current user data and then update the connection repository with a new token.

File: Sources/StreamChat/Repositories/AuthenticationRepository.swift
Comment (martinmitrevski) [inline Sources/StreamChat/Repositories/AuthenticationRepository.swift:280]: This is already handled in the other PR.

**Review by martinmitrevski — APPROVED** [review]
looks really good @polqf 👍 Added few comments, but nothing major. 
Should we have a CHANGELOG entry for this? Also, let's merge it after the patch release.

File: Sources/StreamChat/Repositories/AuthenticationRepository.swift
Comment (polqf) [reply Sources/StreamChat/Repositories/AuthenticationRepository.swift:173]: Oh wow, did not realize 😂 After abstracting the logic, there are no more differences

File: Sources/StreamChat/Repositories/AuthenticationRepository.swift
Comment (polqf) [reply Sources/StreamChat/Repositories/AuthenticationRepository.swift:197]: It looks semantically incorrect. The second method has a completion block, and we should not make assumptions on whether the completion block will be run sync or async. That's why I kept that order.

Another option would be to add this line to the completion of `clearCurrentUserData`, but seems too much for this.

Comment (martinmitrevski) [reply Sources/StreamChat/Repositories/AuthenticationRepository.swift:191]: ok, let's keep it in this order then.

File: Sources/StreamChat/Repositories/ConnectionRepository.swift
Comment (nuno-vieira) [reply Sources/StreamChat/Repositories/ConnectionRepository.swift:14]: I think writing Repositories as concrete classes usually are not the best practice. Repositories by definition are interface adapters, they abstract how we fetch data. By creating concrete classes instead it does not give us the flexibility that repositories should give. With interfaces, we can easily create Decorators, which abstract the Threading code. This will also improve a lot how we test this. 

As abstracting all together thread-safe code, I don't think it will be possible, unless we go back to @Atomic 🤔 Which we want to move away from.

File: Sources/StreamChat/Controllers/ConnectionController/ConnectionController.swift
Comment (nuno-vieira) [inline Sources/StreamChat/Controllers/ConnectionController/ConnectionController.swift:97]: I remember before, we could not have weak-self, otherwise the Updater would not work. This is not the case anymore?

Comment (nuno-vieira) [inline Sources/StreamChat/Controllers/ConnectionController/ConnectionController.swift:97]: I remember before, we could not have weak-self, otherwise, the Updater would not work. This is not the case anymore?

File: Sources/StreamChat/Repositories/ConnectionRepository.swift
Comment (nuno-vieira) [inline Sources/StreamChat/Repositories/ConnectionRepository.swift:94]: Can we actually make this argument without a default value? On the client side, maybe it makes sense, but here in the repository, it makes sense to not omit this decision 🤔

**Review by nuno-vieira — APPROVED** [review]
LGTM! ✅ Just added a couple of NITs, not a blocker at all

File: Sources/StreamChat/Controllers/ConnectionController/ConnectionController.swift
Comment (polqf) [reply Sources/StreamChat/Controllers/ConnectionController/ConnectionController.swift:97]: Tests were failing because we test that the controller is successfully deallocated, so had to change that.
Still, rest of the tests are passing fine, so no issue there.

File: Sources/StreamChat/Repositories/ConnectionRepository.swift
Comment (polqf) [reply Sources/StreamChat/Repositories/ConnectionRepository.swift:94]: Good point

**Pull request merged**
